### PR TITLE
feat(craft): Docker compose proxy path phase 0

### PR DESCRIPTION
**Pull request by acaprau**
## Description

First of 7 branches bringing the sandbox-egress proxy and approval gate (K8s-only today, #11302) to `SANDBOX_BACKEND=docker`. Adds docker-side implementations of two existing `Protocol`s plus a dispatch helper:

- `FileCAStore` (`ca_docker.py`) -- CA over a shared compose volume. `O_EXCL` rendezvous mirrors `K8sSecretCAStore`'s 409-Conflict posture.
- `DockerEventsLookup` (`identity_docker.py`) -- docker-events watcher mirroring the K8s informer (initial-list + watch with `since=<pre-list>`, fail-loud on dup IPs, stop via `CancellableStream.close()`).
- `build_ca_store()` / `build_ip_lookup()` (`backend.py`) -- dispatch on `SANDBOX_BACKEND` with deferred imports.

The proxy's boot path is **not** wired to the dispatch in this PR -- that's branch 1. Code here is dead-but-tested so it can be reviewed in isolation. Zero runtime impact on K8s.

`docs/craft/features/approvals/phase-5-docker.md` is the design doc for the full 7-branch stack.

## How Has This Been Tested?

- [x] `pytest backend/tests/unit/sandbox_proxy/{test_ca_docker,test_identity_docker,test_backend_dispatch}.py` -- 32 passed
- [x] `pre-commit run --all-files` -- clean
- [x] Smoke-tested end-to-end as part of the full stack (proxy container on the sandbox bridge, real CA volume, real sandbox container hitting example.com through the proxy -> 200, identity resolved from container label). Details in design doc `Testing` section.

## Additional Options

- [ ] [Optional] Please cherry-pick this PR to the latest release version.
- [x] [Optional] Override Linear Check

---
## Summary by cubic
Prepares docker-compose backend support for the sandbox proxy by adding backend dispatch helpers, a file-based CA store on a shared volume, and a Docker events–driven IP lookup. Code is dead-but-tested; proxy wiring lands later with no runtime impact.

- New Features
  - Dispatch in `onyx.sandbox_proxy.backend` builds K8s or Docker implementations based on `SANDBOX_BACKEND` with lazy imports.
  - `FileCAStore` persists `ca.crt`/`ca.key` under `SANDBOX_PROXY_CA_VOLUME_PATH` using `O_EXCL` rendezvous and `0o644`/`0o600` modes; fails loud on half-written state.
  - `DockerEventsLookup` initial-lists and watches `DockerClient.events()` filtered by sandbox labels and `SANDBOX_DOCKER_NETWORK`, evicts on `die`/`destroy` and IP changes, reconnects with backoff while clearing sync on disconnect, and rejects duplicate IPs; stop unblocks via `CancellableStream.close()`.
  - Adds Docker label constants `LABEL_DOCKER_COMPONENT`/`LABEL_DOCKER_COMPONENT_SANDBOX` and a fixed-path `SANDBOX_PROXY_CA_VOLUME_PATH`; unit tests cover dispatch, file-store edge cases, event-stream stop/close races, and duplicate-IP behavior. Docs include the compose plan (`phase-5-docker.md`).

- Migration
  - No runtime change in this PR. When wiring lands, compose deployments will need `SANDBOX_BACKEND=docker`, Docker socket read access, and a shared CA volume at `SANDBOX_PROXY_CA_VOLUME_PATH` (proxy RW; sandboxes RO).

<sup>Written for commit 707130f1b72b1edd7f44e3a477f99bf61b041f8f. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/onyx-dot-app/onyx/pull/11507?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: backend/onyx/sandbox_proxy/ca_docker.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+"""Docker-compose implementation of the CA persistence layer.
+
+A shared compose-named volume is the source of truth: the proxy mounts
+it read-write so it can persist on cold start; every sandbox container
+mounts it read-only so ``firewall-init.sh`` can install ``ca.crt`` into
+the trust store.
+
+Cold-start race: two proxy replicas booting against a cold volume both
```
Comment (wenxi-onyx) [inline backend/onyx/sandbox_proxy/ca_docker.py]: Why would there be two replicas of a proxy in docker mode? Is this necessary?

File: backend/onyx/sandbox_proxy/identity_docker.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+from onyx.server.features.build.sandbox.labels import LABEL_TENANT_ID
+from onyx.utils.logger import setup_logger
+
+logger = setup_logger()
+
+_RECONNECT_INITIAL_SECONDS = 1.0
+_RECONNECT_MAX_SECONDS = 30.0
+
+# Mirrors ``LABEL_COMPONENT`` / ``LABEL_COMPONENT_VALUE`` in
+# ``docker_sandbox_manager.py``. Duplicated here rather than imported so
```
Comment (wenxi-onyx) [inline backend/onyx/sandbox_proxy/identity_docker.py]: nit: pref moving the two vars into a shared file that both callers can import without the extra import baggage

File: backend/onyx/server/features/build/configs.py [hunk 3]
```diff
@@ -189,6 +189,13 @@ def _parse_sandbox_backend(raw: str) -> SandboxBackend:
     "SANDBOX_PROXY_CA_CONFIGMAP", "sandbox-proxy-ca-bundle"
 )
 
+# Filesystem path the docker FileCAStore reads/writes the proxy CA at.
+# Both the proxy (RW) and every sandbox container (RO) mount the same
+# named compose volume here. Ignored when SANDBOX_BACKEND=kubernetes.
+SANDBOX_PROXY_CA_VOLUME_PATH = os.environ.get(
```
Comment (wenxi-onyx) [inline backend/onyx/server/features/build/configs.py]: Does this need to be an env var? I think it's fine to just have it as a normal constant?

**Review by wenxi-onyx — APPROVED** [review]

File: backend/onyx/sandbox_proxy/backend.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+
+def build_ca_store() -> CAStore:
+    if SANDBOX_BACKEND is SandboxBackend.KUBERNETES:
+        from onyx.sandbox_proxy.ca_k8s import K8sSecretCAStore
+
+        return K8sSecretCAStore()
+    if SANDBOX_BACKEND is SandboxBackend.DOCKER:
+        from onyx.sandbox_proxy.ca_docker import FileCAStore
+
```
[the remark this answers is not in the document]
Comment (wenxi-onyx) [reply backend/onyx/sandbox_proxy/backend.py:27]: Assuming this is in a follow-up PR?

File: backend/tests/unit/sandbox_proxy/test_backend_dispatch.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        ),
+        patch(
+            "onyx.sandbox_proxy.identity_k8s.K8sInformerLookup",
+            return_value=k8s_ip or MagicMock(),
+        ),
+        patch(
+            "onyx.sandbox_proxy.identity_docker.DockerEventsLookup",
+            return_value=docker_ip or MagicMock(),
+        ),
+    ]
```
[the remark this answers is not in the document]
Comment (wenxi-onyx) [reply backend/tests/unit/sandbox_proxy/test_backend_dispatch.py]: Not valid?

File: backend/onyx/sandbox_proxy/identity_docker.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+            except Exception:
+                self._synced.clear()
+                logger.exception(
+                    "unexpected docker events failure; reconnecting in %.1fs",
+                    backoff,
+                )
+
+            if self._stop_event.wait(backoff):
+                return
+            backoff = min(backoff * 2, _RECONNECT_MAX_SECONDS)
```
[the remark this answers is not in the document]
Comment (wenxi-onyx) [reply backend/onyx/sandbox_proxy/identity_docker.py:191]: Although I don't think this is a real issue, both greptile and codex called it out, so maybe valid lol. This would only be an issue during startup right? I.e. this is still setting up and someone immediately spawns a sandbox?

File: backend/onyx/sandbox_proxy/identity_docker.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+                if self._stop_event.is_set():
+                    return
+                self._apply_event(event)
+        finally:
+            close = getattr(stream, "close", None)
+            if callable(close):
+                try:
+                    close()
+                except Exception:
+                    pass
```
[the remark this answers is not in the document]
Comment (wenxi-onyx) [reply backend/onyx/sandbox_proxy/identity_docker.py]: Maybe valid?

File: backend/onyx/sandbox_proxy/backend.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (acaprau) [reply backend/onyx/sandbox_proxy/backend.py:27]: yes.

File: backend/tests/unit/sandbox_proxy/test_backend_dispatch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (acaprau) [reply backend/tests/unit/sandbox_proxy/test_backend_dispatch.py]: correct. our repo always has k8s and docker imported.

File: backend/onyx/sandbox_proxy/identity_docker.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (acaprau) [reply backend/onyx/sandbox_proxy/identity_docker.py:191]: this was valid feedback. adding a fix + test.

File: backend/onyx/sandbox_proxy/identity_docker.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (acaprau) [reply backend/onyx/sandbox_proxy/identity_docker.py]: yep, valid and fixing.

File: backend/onyx/sandbox_proxy/identity_docker.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        with self._cache_lock:
+            self._cache = new_cache
+            self._by_id = new_by_id
+
+        logger.info(
+            "docker events initial sync: %d sandbox containers cached", len(new_cache)
+        )
+
+    def _watch_loop(self) -> None:
+        stream = self._docker.events(
```
[the remark this answers is not in the document]
Comment (acaprau) [reply backend/onyx/sandbox_proxy/identity_docker.py:230]: resolved in duplicate earlier comment.

File: docs/craft/features/approvals/phase-5-docker.md [hunk 9]
```diff
@@ -0,0 +1,673 @@
...
+an active grant -- the agent cannot run iptables -- but it represents
+a weaker security argument than the K8s init-container model (where
+`NET_ADMIN` is granted only to the init container's security context,
+not to the running sandbox container at all).
+
+Change the script's tail to explicitly clear the Bounding set before
+exec:
+
+```bash
+exec capsh --caps='' --user=craft -- -c 'exec "$@"' -- "$@"
```
[the remark this answers is not in the document]
Comment (acaprau) [reply docs/craft/features/approvals/phase-5-docker.md:364]: Correct on the technical claim — --caps='' doesn't touch the bounding set. This was superseded during smoke testing in #11517: capsh -- invokes /bin/bash and mangles the args for non-script targets, so we pivoted to setpriv --reuid=1000 --regid=1000 --init-groups --bounding-set=-all -- "$@", which (a) execve's the target directly instead of going through bash, and (b) correctly clears the bounding set via --bounding-set=-all.

File: backend/onyx/sandbox_proxy/identity_docker.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+    def _run(self) -> None:
+        backoff = _RECONNECT_INITIAL_SECONDS
+        while not self._stop_event.is_set():
+            try:
+                self._initial_sync()
+                self._initial_sync_done.set()
+                self._synced.set()
+                backoff = _RECONNECT_INITIAL_SECONDS
+                self._watch_loop()
```
[the remark this answers is not in the document]
Comment (acaprau) [reply backend/onyx/sandbox_proxy/identity_docker.py]: valid, `backend/onyx/sandbox_proxy/identity_k8s.py` has this issue too, will fix in a followup.

File: backend/onyx/sandbox_proxy/identity_docker.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+            target=self._run, name="sandbox-proxy-docker-events", daemon=True
+        )
+
+    def start(self) -> None:
+        if self._thread.is_alive():
+            return
+        self._thread.start()
+
+    def stop(self) -> None:
+        self._stop_event.set()
```
[the remark this answers is not in the document]
Comment (acaprau) [reply backend/onyx/sandbox_proxy/identity_docker.py:137]: fixed above.

File: backend/onyx/sandbox_proxy/ca_docker.py (hunk 1 above)
Comment (acaprau) [reply backend/onyx/sandbox_proxy/ca_docker.py]: we would prob not explicitly do this, more of a defense + we have similar guarding logic for the k8s flow. will clarify this in the docstring.

**Comment by acaprau** [comment]
@greptile

**Comment by acaprau** [comment]
@cubic @greptile
